File: auto_test.c

#include "fir_filter.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

void print_usage(const char* prog_name) {
    printf("Usage: %s <numtaps> <fs> <window_type> <cutoff1> [cutoff2 ...]\n", prog_name);
    printf("  numtaps:    Number of filter taps (must be odd)\n");
    printf("  fs:         Sampling frequency in Hz\n");
    printf("  window_type: Window type (number or name):\n");
    printf("               0=Rectangular, 1=Hamming, 2=Blackman, 3=Triangular\n");
    printf("               4=Parzen, 5=Bohman, 6=Nuttall, 7=Blackman-Harris\n");
    printf("               8=Flattop, 9=Bartlett, 10=Hann, 11=Cosine\n");
    printf("  cutoffs:    List of cutoff frequencies in Hz (must be even number of cutoffs)\n");
    printf("\nExample: %s 51 1000.0 1 200.0 300.0\n", prog_name);
    printf("Example: %s 101 44100.0 hann 500.0 1000.0 3000.0 4000.0\n", prog_name);
}

int parse_window_type(const char* str) {
    if (strcmp(str, "0") == 0 || strcasecmp(str, "rectangular") == 0 || strcasecmp(str, "boxcar") == 0) return RECTANGULAR;
    if (strcmp(str, "1") == 0 || strcasecmp(str, "hamming") == 0) return HAMMING;
    if (strcmp(str, "2") == 0 || strcasecmp(str, "blackman") == 0) return BLACKMAN;
    if (strcmp(str, "3") == 0 || strcasecmp(str, "triangular") == 0) return TRIANGULAR;
    if (strcmp(str, "4") == 0 || strcasecmp(str, "parzen") == 0) return PARZEN;
    if (strcmp(str, "5") == 0 || strcasecmp(str, "bohman") == 0) return BOHMAN;
    if (strcmp(str, "6") == 0 || strcasecmp(str, "nuttall") == 0) return NUTTALL;
    if (strcmp(str, "7") == 0 || strcasecmp(str, "blackmanharris") == 0 || strcasecmp(str, "blackman-harris") == 0) return BLACKMANHARRIS;
    if (strcmp(str, "8") == 0 || strcasecmp(str, "flattop") == 0) return FLATTOP;
    if (strcmp(str, "9") == 0 || strcasecmp(str, "bartlett") == 0) return BARTLETT;
    if (strcasecmp(str, "10") == 0 || strcasecmp(str, "hann") == 0) return HANN;
    if (strcasecmp(str, "11") == 0 || strcasecmp(str, "cosine") == 0) return COSINE;
    return -1;
}

const char* get_window_name(enum fir_filter_window_type window) {
    switch (window) {
        case RECTANGULAR:   return "Rectangular (boxcar)";
        case HAMMING:       return "Hamming";
        case BLACKMAN:      return "Blackman";
        case TRIANGULAR:    return "Triangular";
        case PARZEN:        return "Parzen";
        case BOHMAN:        return "Bohman";
        case NUTTALL:       return "Nuttall";
        case BLACKMANHARRIS:return "Blackman-Harris";
        case FLATTOP:       return "Flat-top";
        case BARTLETT:      return "Bartlett";
        case HANN:          return "Hann";
        case COSINE:        return "Cosine (sine)";
        default:            return "Unknown";
    }
}

int main(int argc, char* argv[]) {
    // Check minimum number of arguments
    if (argc < 5) {
        print_usage(argv[0]);
        return 1;
    }
    
    // Parse command line arguments
    int numtaps = atoi(argv[1]);
    float fs = (float)atof(argv[2]);
    int window_type = parse_window_type(argv[3]);
    
    if (window_type == -1) {
        fprintf(stderr, "Error: Invalid window type. Must be 0, 1, or 2.\n");
        return 1;
    }
    
    int num_cutoffs = argc - 4;
    if (num_cutoffs % 2 != 0) {
        fprintf(stderr, "Error: Number of cutoff frequencies must be even.\n");
        return 1;
    }
    
    float* cutoffs = (float*)malloc(num_cutoffs * sizeof(float));
    if (!cutoffs) {
        fprintf(stderr, "Error: Memory allocation failed.\n");
        return 1;
    }
    
    for (int i = 0; i < num_cutoffs; i++) {
        cutoffs[i] = (float)atof(argv[4 + i]);
    }
    
    // Allocate array for coefficients
    float* h = (float*)malloc(numtaps * sizeof(float));
    if (!h) {
        fprintf(stderr, "Error: Memory allocation failed.\n");
        free(cutoffs);
        return 1;
    }
    
    // Design the filter
    printf("# FIR Filter Design\n");
    printf("# Taps: %d\n", numtaps);
    printf("# Sampling frequency: %.1f Hz\n", fs);
    printf("# Window: %s\n", get_window_name(window_type));
    printf("# Cutoffs: ");
    for (int i = 0; i < num_cutoffs; i++) {
        printf("%.1f ", cutoffs[i]);
    }
    printf("Hz\n\n");
    
    int result = firwin(numtaps, num_cutoffs, cutoffs, fs, window_type, h);
    if (result != 0) {
        fprintf(stderr, "Error designing filter.\n");
        free(cutoffs);
        free(h);
        return 1;
    }
    
    // Print all coefficients
    printf("# Coefficients:\n");
    for (int i = 0; i < numtaps; i++) {
        printf("%.15g\n", h[i]);
    }
    
    // Calculate and print the sum of coefficients
    float sum = 0.0f;
    for (int i = 0; i < numtaps; i++) {
        sum += h[i];
    }
    printf("\n# Sum of coefficients: %.15g\n", sum);
    
    free(cutoffs);
    free(h);
    return 0;
}

File: auto_test.py
#!/usr/bin/env python3

import numpy as np
import subprocess
import random
import matplotlib.pyplot as plt
from scipy import signal
import os

# Compile the C code if needed
if not os.path.exists('auto_test'):
    print("Compiling C code...")
    subprocess.run(['make', 'clean'], check=True)
    subprocess.run(['make'], check=True)

def run_c_firwin(numtaps, fs, window_type, cutoffs):
    """Run the C implementation and return the coefficients."""
    # Prepare the command line arguments
    args = ['./auto_test', str(numtaps), str(fs), str(window_type)] + [str(c) for c in cutoffs]
    
    # Run the command and capture output
    result = subprocess.run(args, capture_output=True, text=True)
    
    if result.returncode != 0:
        print("Error running C implementation:")
        print(result.stderr)
        return None
    
    # Parse the output to get coefficients
    coefficients = []
    in_coeffs = False
    
    for line in result.stdout.split('\n'):
        line = line.strip()
        if line.startswith('#'):
            if 'Coefficients:' in line:
                in_coeffs = True
            continue
        if in_coeffs and line:
            try:
                coeff = float(line)
                coefficients.append(coeff)
            except ValueError:
                pass
    
    return np.array(coefficients)

def run_scipy_firwin(numtaps, fs, window_type, cutoffs):
    """Run the SciPy implementation and return the coefficients."""
    # Map window type to SciPy window names and parameters
    window_info = {
        0: ('boxcar', {}),                # Rectangular
        1: ('hamming', {}),               # Hamming
        2: ('blackman', {}),              # Blackman
        3: ('triang', {}),                # Triangular
        4: ('boxcar', {}),                # Parzen (not in scipy, using boxcar as fallback)
        5: ('bohman', {}),                # Bohman
        6: ('nuttall', {}),               # Nuttall
        7: ('blackmanharris', {}),        # Blackman-Harris
        8: ('flattop', {}),               # Flat-top
        9: ('bartlett', {}),              # Bartlett
        10: ('hann', {}),                 # Hann
        11: ('cosine', {})                # Cosine (sine)
    }
    
    
    # Handle the case where the first cutoff is 0 (lowpass/bandpass)
    pass_zero = cutoffs[0] == 0
    pass_nyquist = cutoffs[-1] >= fs/2
    if pass_zero:
        cutoffs = cutoffs[1:]
    if pass_nyquist:
        cutoffs = cutoffs[:-1]
    
    # Get window type and parameters
    window_name, window_params = window_info[window_type]
    
    # Call SciPy's firwin with appropriate parameters

    coeffs = signal.firwin(
        numtaps,
        cutoffs,
        window=window_name,
        pass_zero=pass_zero,
        fs=fs,
        **window_params
    )
    
    # For cosine window, we need to use a different approach as scipy doesn't have it directly
    if window_type == 11:  # Cosine window
        n = np.arange(numtaps)
        coeffs = np.sin(np.pi * (n + 0.5) / numtaps)
    
    return coeffs



if __name__ == "__main__":
    PLOT_RESULTS = False
    TEST_NUM = 1000
    PRINT_DEBUG = False

    # Test each window type with a simple low-pass filter
    window_names = [
        'Rectangular', 'Hamming', 'Blackman', 'Triangular',
        'Parzen', 'Bohman', 'Nuttall', 'Blackman-Harris',
        'Flat-top', 'Bartlett', 'Hann', 'Cosine'
    ]
    for window_type in range(12):
        print(f"\nTesting {window_names[window_type]} window")
        print("=" * 50)

        test_results = []

        for test_idx in range(TEST_NUM):
            # generate random params
            fs = random.randint(1000, 5000)
            numtaps = random.randint(5, 100)
            num_bands = random.randint(1, 10)
            cutoffs = [random.random() * (fs//2-2) + 1 for _ in range(num_bands * 2)]
            cutoffs.sort()
            pass_zero = random.choice([True, False])
            pass_nyquist = random.choice([True, False])
            if numtaps % 2 == 0:
                pass_nyquist = False
            
            if num_bands == 1 and pass_zero and pass_nyquist:
                if random.choice([True, False]):
                    pass_zero = False
                else:
                    pass_nyquist = False

            if pass_zero:
                cutoffs[0] = 0
            if pass_nyquist:
                cutoffs[-1] = fs/2

            if PRINT_DEBUG:
                print(f"Test {test_idx + 1}: fs={fs}, numtaps={numtaps}, num_bands={num_bands}, cutoffs={cutoffs}, pass_zero={pass_zero}, pass_nyquist={pass_nyquist}")

            # Run both implementations
            c_coeffs = run_c_firwin(numtaps, fs, window_type, cutoffs)
            py_coeffs = run_scipy_firwin(numtaps, fs, window_type, cutoffs)
        
            if PLOT_RESULTS:
                # Plot results
                plt.figure(figsize=(12, 6))
                
                # Time domain
                plt.subplot(1, 2, 1)
                plt.plot(c_coeffs, 'b-', linewidth=2, label='C Implementation')
                plt.plot(py_coeffs, 'r--', linewidth=1.5, label='SciPy')
                plt.title(f'{window_names[window_type]} Window - Time Domain')
                plt.xlabel('Sample')
                plt.ylabel('Amplitude')
                plt.legend()
                plt.grid(True)
                
                # Frequency domain
                plt.subplot(1, 2, 2)
                for coeffs, color, label in [(c_coeffs, 'b-', 'C'), (py_coeffs, 'r--', 'SciPy')]:
                    w, h = signal.freqz(coeffs, fs=fs)
                    plt.semilogx(w, 20 * np.log10(np.maximum(np.abs(h), 1e-10)), 
                                color, linewidth=2, label=f'{label} Response')
                
                plt.title(f'{window_names[window_type]} Window - Frequency Response')
                plt.xlabel('Frequency [Hz]')
                plt.ylabel('Magnitude [dB]')
                plt.ylim(-100, 10)
                plt.xlim(1, fs/2)
                for cutoff in cutoffs:
                    plt.axvline(cutoff, color='k', linestyle='--', alpha=0.7)
                plt.axhline(-3, color='k', linestyle='--', alpha=0.7)
                plt.legend()
                plt.grid(True, which='both', axis='both')
                
                plt.tight_layout()
                plt.show()

                break

            else:
                # check if the two arrays are equal
                if np.allclose(c_coeffs, py_coeffs, rtol=1e-4, atol=1e-4):
                    test_results.append(1)
                    if PRINT_DEBUG:
                        print(f"Test {test_idx + 1} passed")
                else:
                    test_results.append(0)
                    if PRINT_DEBUG:
                        print(f"Test {test_idx + 1} failed")
            
        print(f"Pass rate: {sum(test_results) / len(test_results)}")
                    
